lib/Driver/Driver.cpp

```diff
@@ -1705,12 +1723,13 @@ Driver::buildOutputFileMap(const llvm::opt::DerivedArgList &Args,
     return nullptr;
 
   // TODO: perform some preflight checks to ensure the file exists.
-  auto OFM = OutputFileMap::loadFromPath(A->getValue(), workingDirectory);
-  if (!OFM) {
-    // TODO: emit diagnostic with error string
-    Diags.diagnose(SourceLoc(), diag::error_unable_to_load_output_file_map);
+  llvm::Expected<OutputFileMap> OFM =
+      OutputFileMap::loadFromPath(A->getValue(), workingDirectory);
+  if (auto Err = OFM.takeError()) {
+    Diags.diagnose(SourceLoc(), diag::error_unable_to_load_output_file_map,
+                   llvm::toString(std::move(Err)));
   }
-  return OFM;
+  return llvm::make_unique<OutputFileMap>(*OFM);
```

Comment:
Uh, hm. I guess if this is properly moving the OFM it's okay (i.e. we won't copy the whole hash map). Can you use `std::move(*OFM)` here?

Reply:
(This is probably one of those things that doesn't matter in practice, but it is something I noticed.)

Reply:
I've changed it to return an Optional<OutputFileMap>, but I'm not sure where I need the move's and where the compiler will do the optimization for me (partly because there's an implicit conversion to Optional).

Reply:
The compiler will only do the optimization if it's the direct "argument" of `return`; pretty much never else. (I think there's a warning now for if you *do* use `std::move` when it's unnecessary, so you can err on the side of moving when there's a type big enough to warrant it.)